compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -212,8 +226,8 @@ fn mir_build(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_
 
             let return_ty = typeck_results.node_type(id);
 
-            let ast_expr = &tcx.hir().body(body_id).value;
-            let (thir, expr) = build_thir(tcx, def, ast_expr);
+            let (thir, expr) = tcx.thir_body(def);
+            let thir = thir.steal();
```

Comment:
for this to work, we need to invoke the "check unsafety" query first, right? how do we make sure that we do this?

Reply:
I do that a few lines above, at the beginning of `mir_build`: https://github.com/rust-lang/rust/blob/abaf435e2379a1e66ab6e58b1daa87042dba28bc/compiler/rustc_mir_build/src/build/mod.rs#L49-L59

Reply:
I see. Can we add a comment about that, or perhaps move the steal closer together?

Reply:
Mhh, by the way I think we want to steal the THIR even if a type error occurs, so we can free the memory, which is not what happens right now.

Reply:
yes, that makes sense. I was thinking also that we should compare memory usage